File: transports/csharp/tests/Csilgen.Transport.Tests/ConformanceTests.cs
// Verify the C# reference library against the checked-in conformance vectors.
//
// This both guards the C# encoders/decoders against drift and demonstrates the contract every
// reference library follows: reconstruct each vector's envelope from its language-neutral `input`,
// assert encode -> `hex`, and assert decode(`hex`) -> the same envelope. Ported from
// transports/rust/tests/conformance.rs and transports/go/conformance_test.go.

using System.Text.Json;
using Csilgen.Transport;
using Xunit;

namespace Csilgen.Transport.Tests;

public sealed class ConformanceTests
{
    // Walk up from the test assembly's base directory to find transports/conformance, mirroring how
    // the Go test resolves "../conformance" relative to its source. Robust to bin/Debug/net8.0 depth.
    private static string ConformanceDir()
    {
        DirectoryInfo? dir = new(AppContext.BaseDirectory);
        while (dir is not null)
        {
            string nested = Path.Combine(dir.FullName, "transports", "conformance");
            if (File.Exists(Path.Combine(nested, "rpc.json")))
            {
                return nested;
            }

            string sibling = Path.Combine(dir.FullName, "conformance");
            if (File.Exists(Path.Combine(sibling, "rpc.json")))
            {
                return sibling;
            }

            dir = dir.Parent;
        }

        throw new DirectoryNotFoundException("could not locate transports/conformance");
    }

    private static IReadOnlyList<JsonElement> LoadVectors(string name)
    {
        string path = Path.Combine(ConformanceDir(), name);
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        var vectors = new List<JsonElement>();
        foreach (JsonElement v in doc.RootElement.GetProperty("vectors").EnumerateArray())
        {
            // Clone so the elements outlive the JsonDocument's disposal.
            vectors.Add(v.Clone());
        }

        return vectors;
    }

    private static byte[] Unhex(string s)
    {
        if (s.Length == 0)
        {
            return Array.Empty<byte>();
        }

        byte[] result = new byte[s.Length / 2];
        for (int i = 0; i < result.Length; i++)
        {
            result[i] = Convert.ToByte(s.Substring(i * 2, 2), 16);
        }

        return result;
    }

    private static string Hex(byte[] bytes) => Convert.ToHexString(bytes).ToLowerInvariant();

    private static string ReqStr(JsonElement input, string key) => input.GetProperty(key).GetString()!;

    private static ulong ReqU64(JsonElement input, string key) => input.GetProperty(key).GetUInt64();

    private static long ReqI64(JsonElement input, string key) => input.GetProperty(key).GetInt64();

    private static string? OptStr(JsonElement input, string key) =>
        input.TryGetProperty(key, out JsonElement e) && e.ValueKind != JsonValueKind.Null
            ? e.GetString()
            : null;

    private static ulong? OptU64(JsonElement input, string key) =>
        input.TryGetProperty(key, out JsonElement e) && e.ValueKind != JsonValueKind.Null
            ? e.GetUInt64()
            : null;

    [Fact]
    public void RpcVectors()
    {
        foreach (JsonElement vec in LoadVectors("rpc.json"))
        {
            string name = vec.GetProperty("name").GetString()!;
            string expectedHex = vec.GetProperty("hex").GetString()!;
            JsonElement input = vec.GetProperty("input");
            byte[] actual;

            switch (ReqStr(input, "kind"))
            {
                case "request":
                    {
                        var r = RpcRequest.New(
                            ReqStr(input, "service"),
                            ReqStr(input, "op"),
                            Unhex(ReqStr(input, "payload_hex"))) with
                        {
                            Id = OptU64(input, "id"),
                            Auth = OptStr(input, "auth"),
                        };
                        actual = r.Encode();
                        RpcRequest dec = RpcRequest.Decode(actual);
                        Assert.Equal(r.Service, dec.Service);
                        Assert.Equal(r.Op, dec.Op);
                        Assert.Equal(r.Id, dec.Id);
                        Assert.Equal(r.Auth, dec.Auth);
                        Assert.True(r.Payload.AsSpan().SequenceEqual(dec.Payload), $"payload {name}");
                        break;
                    }

                case "response":
                    {
                        var r = new RpcResponse(
                            Status.FromCode((int)ReqI64(input, "status")),
                            Unhex(ReqStr(input, "payload_hex")))
                        {
                            Id = OptU64(input, "id"),
                            Variant = OptStr(input, "variant"),
                            Error = OptStr(input, "error"),
                        };
                        actual = r.Encode();
                        RpcResponse dec = RpcResponse.Decode(actual);
                        Assert.Equal(r.Id, dec.Id);
                        Assert.Equal(r.Status, dec.Status);
                        Assert.Equal(r.Variant, dec.Variant);
                        Assert.Equal(r.Error, dec.Error);
                        Assert.True(r.Payload.AsSpan().SequenceEqual(dec.Payload), $"payload {name}");
                        break;
                    }

                case "push":
                    {
                        var p = RpcPush.New(
                            ReqStr(input, "service"),
                            ReqStr(input, "event"),
                            Unhex(ReqStr(input, "payload_hex")));
                        actual = p.Encode();
                        RpcPush dec = RpcPush.Decode(actual);
                        Assert.Equal(p.Service, dec.Service);
                        Assert.Equal(p.Event, dec.Event);
                        Assert.True(p.Payload.AsSpan().SequenceEqual(dec.Payload), $"payload {name}");
                        break;
                    }

                default:
                    throw new Xunit.Sdk.XunitException($"unknown rpc kind in {name}");
            }

            Assert.Equal(expectedHex, Hex(actual));
        }
    }

    [Fact]
    public void EventsVectors()
    {
        foreach (JsonElement vec in LoadVectors("events.json"))
        {
            string name = vec.GetProperty("name").GetString()!;
            string expectedHex = vec.GetProperty("hex").GetString()!;
            JsonElement input = vec.GetProperty("input");

            // Control payloads are keyed by "control"; application events by "profile".
            if (OptStr(input, "control") is string control)
            {
                byte[] controlBytes = control switch
                {
                    "hello" => new Hello(
                            ReadU64Array(input, "versions"),
                            ReadStrArray(input, "profiles"))
                    {
                        Service = OptStr(input, "service"),
                        Auth = OptStr(input, "auth"),
                    }.Encode(),
                    "hello_ack" => new HelloAck(ReqU64(input, "v"), ReqStr(input, "profile"))
                    {
                        Session = OptStr(input, "session"),
                    }.Encode(),
                    "ping" => new Heartbeat(ReqU64(input, "nonce"))
                    {
                        At = OptU64(input, "at"),
                    }.Encode(),
                    "close" => new Close(Status.FromCode((int)ReqI64(input, "status")))
                    {
                        Reason = OptStr(input, "reason"),
                    }.Encode(),
                    _ => throw new Xunit.Sdk.XunitException($"unknown control in {name}"),
                };
                Assert.Equal(expectedHex, Hex(controlBytes));
                continue;
            }

            string profileStr = ReqStr(input, "profile");
            Assert.True(ProfileExtensions.TryParse(profileStr, out Profile profile), name);
            byte[] payload = Unhex(ReqStr(input, "payload_hex"));
            Event ev = profile switch
            {
                Profile.Verbose => Event.Verbose(
                    OptStr(input, "service"), ReqStr(input, "event"), payload) with
                {
                    Id = OptU64(input, "id"),
                },
                Profile.Compact => Event.Compact(
                    ReqU64(input, "service_ord"), ReqU64(input, "op_ord"), payload) with
                {
                    Id = OptU64(input, "id"),
                },
                _ => throw new Xunit.Sdk.XunitException($"unknown profile in {name}"),
            };
            byte[] bytes = ev.Encode(profile);
            Assert.Equal(expectedHex, Hex(bytes));

            Event dec = Event.Decode(bytes, profile);
            Assert.Equal(ev.Service, dec.Service);
            Assert.Equal(ev.ServiceOrd, dec.ServiceOrd);
            Assert.Equal(ev.EventName, dec.EventName);
            Assert.Equal(ev.OpOrd, dec.OpOrd);
            Assert.Equal(ev.Id, dec.Id);
            Assert.True(ev.Payload.AsSpan().SequenceEqual(dec.Payload), $"payload {name}");
        }
    }

    [Fact]
    public void DatagramVectors()
    {
        foreach (JsonElement vec in LoadVectors("datagrams.json"))
        {
            string name = vec.GetProperty("name").GetString()!;
            string expectedHex = vec.GetProperty("hex").GetString()!;
            JsonElement input = vec.GetProperty("input");

            switch (ReqStr(input, "profile"))
            {
                case "cbor-array":
                    {
                        var d = Datagram.New(
                            ReqU64(input, "op_ord"),
                            ReqU64(input, "seq"),
                            Unhex(ReqStr(input, "payload_hex")));
                        byte[] bytes = d.Encode();
                        Assert.Equal(expectedHex, Hex(bytes));
                        Datagram dec = Datagram.Decode(bytes);
                        Assert.Equal(d.OpOrd, dec.OpOrd);
                        Assert.Equal(d.Seq, dec.Seq);
                        Assert.True(d.Payload.AsSpan().SequenceEqual(dec.Payload), $"payload {name}");
                        break;
                    }

                case "compact-header":
                    {
                        var d = CompactDatagram.New(
                            (byte)ReqU64(input, "op_ord"),
                            (ushort)ReqU64(input, "seq"),
                            Unhex(ReqStr(input, "body_hex")));
                        if (OptU64(input, "epoch") is ulong epoch)
                        {
                            d = d.WithEpoch((byte)epoch);
                        }

                        byte[] bytes = d.Encode();
                        Assert.Equal(expectedHex, Hex(bytes));
                        CompactDatagram dec = CompactDatagram.Decode(bytes);
                        Assert.Equal(d.OpOrd, dec.OpOrd);
                        Assert.Equal(d.Seq, dec.Seq);
                        Assert.Equal(d.Epoch, dec.Epoch);
                        Assert.True(d.Body.AsSpan().SequenceEqual(dec.Body), $"body {name}");
                        break;
                    }

                default:
                    throw new Xunit.Sdk.XunitException($"unknown datagram profile in {name}");
            }
        }
    }

    private static List<ulong> ReadU64Array(JsonElement input, string key)
    {
        var result = new List<ulong>();
        foreach (JsonElement x in input.GetProperty(key).EnumerateArray())
        {
            result.Add(x.GetUInt64());
        }

        return result;
    }

    private static List<string> ReadStrArray(JsonElement input, string key)
    {
        var result = new List<string>();
        foreach (JsonElement x in input.GetProperty(key).EnumerateArray())
        {
            result.Add(x.GetString()!);
        }

        return result;
    }
}

File: transports/swift/Tests/CsilgenTransportTests/ConformanceTests.swift
// Verify the Swift reference library against the checked-in conformance vectors
// (transports/conformance/*.json). This both guards the encoders/decoders against drift
// and demonstrates the contract every reference library follows: reconstruct each
// vector's envelope from its language-neutral `input`, assert encode → `hex`, and assert
// decode(`hex`) → the same envelope. Ported from transports/go/conformance_test.go.
//
// Foundation is imported only here, in the test target, for JSONSerialization + path
// math — the library itself stays Foundation-free so the wire codec is portable.
import Foundation
import XCTest

@testable import CsilgenTransport

final class ConformanceTests: XCTestCase {
    func testRpcVectors() throws {
        for vec in try loadVectors("rpc.json") {
            let input = vec.input
            let actual: [UInt8]
            switch try reqStr(input, "kind") {
            case "request":
                var r = RpcRequest(
                    service: try reqStr(input, "service"),
                    op: try reqStr(input, "op"),
                    payload: try reqHex(input, "payload_hex"))
                r.id = optU64(input, "id")
                r.auth = optStr(input, "auth")
                XCTAssertEqual(try RpcRequest.decode(r.encode()), r, "\(vec.name) decode")
                actual = r.encode()
            case "response":
                let r = RpcResponse(
                    status: Status(code: try reqInt(input, "status")),
                    payload: try reqHex(input, "payload_hex"),
                    id: optU64(input, "id"),
                    variant: optStr(input, "variant"),
                    error: optStr(input, "error"))
                XCTAssertEqual(try RpcResponse.decode(r.encode()), r, "\(vec.name) decode")
                actual = r.encode()
            case "push":
                let p = RpcPush(
                    service: try reqStr(input, "service"),
                    event: try reqStr(input, "event"),
                    payload: try reqHex(input, "payload_hex"))
                XCTAssertEqual(try RpcPush.decode(p.encode()), p, "\(vec.name) decode")
                actual = p.encode()
            default:
                XCTFail("unknown rpc kind in \(vec.name)")
                continue
            }
            XCTAssertEqual(toHex(actual), vec.hex, "\(vec.name) encode")
        }
    }

    func testEventsVectors() throws {
        for vec in try loadVectors("events.json") {
            let input = vec.input
            // Control payloads are keyed by "control"; application events by "profile".
            if let control = optStr(input, "control") {
                let bytes: [UInt8]
                switch control {
                case "hello":
                    bytes = Hello(
                        versions: try uintArray(input, "versions"),
                        profiles: try stringArray(input, "profiles"),
                        service: optStr(input, "service"),
                        auth: optStr(input, "auth")
                    ).encode()
                case "hello_ack":
                    bytes = HelloAck(
                        v: try reqU64(input, "v"),
                        profile: try reqStr(input, "profile"),
                        session: optStr(input, "session")
                    ).encode()
                case "ping":
                    bytes = Heartbeat(nonce: try reqU64(input, "nonce"), at: optU64(input, "at"))
                        .encode()
                case "close":
                    bytes = Close(
                        status: Status(code: try reqInt(input, "status")),
                        reason: optStr(input, "reason")
                    ).encode()
                default:
                    XCTFail("unknown control \(control)")
                    continue
                }
                XCTAssertEqual(toHex(bytes), vec.hex, "\(vec.name) encode")
                continue
            }

            guard let profile = Profile.parse(try reqStr(input, "profile")) else {
                XCTFail("unknown profile in \(vec.name)")
                continue
            }
            let payload = try reqHex(input, "payload_hex")
            var ev: Event
            switch profile {
            case .verbose:
                ev = Event.verbose(
                    service: optStr(input, "service"), event: try reqStr(input, "event"),
                    payload: payload)
                ev.id = optU64(input, "id")
            case .compact:
                ev = Event.compact(
                    serviceOrd: try reqU64(input, "service_ord"),
                    opOrd: try reqU64(input, "op_ord"), payload: payload)
                ev.id = optU64(input, "id")
            }
            XCTAssertEqual(toHex(try ev.encode(profile)), vec.hex, "\(vec.name) encode")
            XCTAssertEqual(try Event.decode(ev.encode(profile), profile), ev, "\(vec.name) decode")
        }
    }

    func testDatagramVectors() throws {
        for vec in try loadVectors("datagrams.json") {
            let input = vec.input
            switch try reqStr(input, "profile") {
            case "cbor-array":
                let d = Datagram(
                    opOrd: try reqU64(input, "op_ord"), seq: try reqU64(input, "seq"),
                    payload: try reqHex(input, "payload_hex"))
                XCTAssertEqual(toHex(d.encode()), vec.hex, "\(vec.name) encode")
                XCTAssertEqual(try Datagram.decode(d.encode()), d, "\(vec.name) decode")
            case "compact-header":
                var d = CompactDatagram(
                    opOrd: UInt8(try reqU64(input, "op_ord")),
                    seq: UInt16(try reqU64(input, "seq")),
                    body: try reqHex(input, "body_hex"))
                if let e = optU64(input, "epoch") { d = d.withEpoch(UInt8(e)) }
                XCTAssertEqual(toHex(d.encode()), vec.hex, "\(vec.name) encode")
                XCTAssertEqual(try CompactDatagram.decode(d.encode()), d, "\(vec.name) decode")
            default:
                XCTFail("unknown datagram profile in \(vec.name)")
            }
        }
    }
}

// MARK: - Vector loading + language-neutral input helpers

struct Vector {
    let name: String
    let hex: String
    let input: [String: Any]
}

/// The conformance vectors live at <repo>/transports/conformance, outside this package.
/// Walk up from this test source file rather than depend on the working directory.
private func conformanceDir() -> URL {
    var url = URL(fileURLWithPath: #filePath)
    // .../transports/swift/Tests/CsilgenTransportTests/ConformanceTests.swift → up to transports
    for _ in 0..<4 { url.deleteLastPathComponent() }
    return url.appendingPathComponent("conformance")
}

func loadVectors(_ file: String) throws -> [Vector] {
    let path = conformanceDir().appendingPathComponent(file)
    let data = try Data(contentsOf: path)
    // Cast element-by-element rather than to `[[String: Any]]` in one shot: Linux's
    // swift-corelibs-foundation is stricter about bridging nested heterogeneous arrays.
    guard let doc = try JSONSerialization.jsonObject(with: data) as? [String: Any],
        let raw = doc["vectors"] as? [Any]
    else {
        throw TransportError.malformed("vectors file \(file) is not the expected shape")
    }
    return raw.compactMap { entry in
        guard let e = entry as? [String: Any],
            let name = e["name"] as? String, let hex = e["hex"] as? String,
            let input = e["input"] as? [String: Any]
        else { return nil }
        return Vector(name: name, hex: hex, input: input)
    }
}

func toHex(_ bytes: [UInt8]) -> String {
    bytes.map { String(format: "%02x", $0) }.joined()
}

func fromHex(_ s: String) -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(s.count / 2)
    var it = s.startIndex
    while it < s.endIndex {
        let next = s.index(it, offsetBy: 2)
        out.append(UInt8(s[it..<next], radix: 16)!)
        it = next
    }
    return out
}

private func numberValue(_ any: Any?) -> Int64? {
    switch any {
    case let n as NSNumber: return n.int64Value
    case let i as Int: return Int64(i)
    case let d as Double: return Int64(d)
    default: return nil
    }
}

func optStr(_ m: [String: Any], _ k: String) -> String? {
    guard let v = m[k], !(v is NSNull) else { return nil }
    return v as? String
}

func optU64(_ m: [String: Any], _ k: String) -> UInt64? {
    guard let v = m[k], !(v is NSNull), let n = numberValue(v) else { return nil }
    return UInt64(n)
}

func reqStr(_ m: [String: Any], _ k: String) throws -> String {
    guard let s = optStr(m, k) else { throw TransportError.malformed("missing string \(k)") }
    return s
}

func reqU64(_ m: [String: Any], _ k: String) throws -> UInt64 {
    guard let n = optU64(m, k) else { throw TransportError.malformed("missing integer \(k)") }
    return n
}

func reqInt(_ m: [String: Any], _ k: String) throws -> Int {
    guard let v = m[k], let n = numberValue(v) else {
        throw TransportError.malformed("missing integer \(k)")
    }
    return Int(n)
}

func reqHex(_ m: [String: Any], _ k: String) throws -> [UInt8] {
    fromHex(try reqStr(m, k))
}

func uintArray(_ m: [String: Any], _ k: String) throws -> [UInt64] {
    guard let arr = m[k] as? [Any] else { throw TransportError.malformed("missing array \(k)") }
    return arr.compactMap { numberValue($0).map(UInt64.init) }
}

func stringArray(_ m: [String: Any], _ k: String) throws -> [String] {
    guard let arr = m[k] as? [Any] else { throw TransportError.malformed("missing array \(k)") }
    return arr.compactMap { $0 as? String }
}
